File: controller/src/f1tenth_dynamic_mpcc/src/ackermann_command_gate.cpp
#include <ackermann_msgs/AckermannDrive.h>
#include <ackermann_msgs/AckermannDriveStamped.h>
#include <ros/ros.h>
#include <std_msgs/Bool.h>
#include <std_srvs/SetBool.h>

#include <mutex>
#include <optional>

class CommandGate {
 public:
  CommandGate():private_("~"){
    bool allowed=false;private_.param("allow_real_hardware",allowed,false);if(!allowed)throw std::runtime_error("allow_real_hardware must be true");
    private_.param("input_topic",input_,std::string("/f1tenth_mpcc/real/ackermann_cmd_stamped"));private_.param("output_topic",output_,std::string("/tianracer/ackermann_cmd"));private_.param("collision_topic",collision_topic_,std::string("/tianracer/collision"));private_.param("collision_required",collision_required_,false);private_.param("start_enabled",enabled_,false);private_.param("command_timeout_s",command_timeout_,.20);private_.param("collision_timeout_s",collision_timeout_,.50);double rate;private_.param("publish_rate_hz",rate,20.0);
    publisher_=node_.advertise<ackermann_msgs::AckermannDrive>(output_,1);command_sub_=node_.subscribe(input_,1,&CommandGate::commandCallback,this,ros::TransportHints().tcpNoDelay());if(collision_required_)collision_sub_=node_.subscribe(collision_topic_,1,&CommandGate::collisionCallback,this);service_=private_.advertiseService("set_enabled",&CommandGate::enable,this);timer_=node_.createTimer(ros::Duration(1/rate),&CommandGate::tick,this);stop();ROS_WARN("C++ hardware command gate output=%s enabled=%s",output_.c_str(),enabled_?"true":"false");
  }
  ~CommandGate(){stop();}
 private:
  void commandCallback(const ackermann_msgs::AckermannDriveStamped::ConstPtr& msg){std::lock_guard<std::mutex>l(mutex_);command_=msg->drive;command_stamp_=ros::WallTime::now().toSec();}
  void collisionCallback(const std_msgs::Bool::ConstPtr& msg){std::lock_guard<std::mutex>l(mutex_);collision_=msg->data;collision_stamp_=ros::WallTime::now().toSec();}
  bool safe(double now)const{return command_&&now-command_stamp_<=command_timeout_&&(!collision_required_||(!collision_&&now-collision_stamp_<=collision_timeout_));}
  bool enable(std_srvs::SetBool::Request&req,std_srvs::SetBool::Response&res){std::lock_guard<std::mutex>l(mutex_);if(req.data&&!safe(ros::WallTime::now().toSec())){res.success=false;res.message="fresh command required";return true;}enabled_=req.data;res.success=true;res.message=enabled_?"enabled":"disabled";return true;}
  void tick(const ros::TimerEvent&){std::optional<ackermann_msgs::AckermannDrive>command;{std::lock_guard<std::mutex>l(mutex_);if(enabled_&&safe(ros::WallTime::now().toSec()))command=command_;}if(command)publisher_.publish(*command);else stop();}
  void stop(){publisher_.publish(ackermann_msgs::AckermannDrive());}
  ros::NodeHandle node_,private_;ros::Publisher publisher_;ros::Subscriber command_sub_,collision_sub_;ros::ServiceServer service_;ros::Timer timer_;mutable std::mutex mutex_;std::optional<ackermann_msgs::AckermannDrive>command_;double command_stamp_{-1e99},collision_stamp_{-1e99},command_timeout_{},collision_timeout_{};bool collision_{true},collision_required_{false},enabled_{false};std::string input_,output_,collision_topic_;
};

int main(int argc,char**argv){ros::init(argc,argv,"f1tenth_mpcc_command_gate");try{CommandGate gate;ros::spin();}catch(const std::exception&e){ROS_FATAL("command gate failed: %s",e.what());return 1;}return 0;}

File: controller/src/f1tenth_dynamic_mpcc/scripts/ackermann_command_gate.py
#!/usr/bin/env python3
"""Explicitly armed, fail-closed adapter from MPCC Stamped commands to hardware."""

from __future__ import annotations

import threading
import time

import rospy
from ackermann_msgs.msg import AckermannDrive, AckermannDriveStamped
from std_msgs.msg import Bool
from std_srvs.srv import SetBool, SetBoolResponse


class CommandGate:
    def __init__(self) -> None:
        if not bool(rospy.get_param("~allow_real_hardware", False)):
            raise RuntimeError("allow_real_hardware must be explicitly true")
        self.input_topic = str(rospy.get_param("~input_topic", "/f1tenth_mpcc/real/ackermann_cmd_stamped"))
        self.output_topic = str(rospy.get_param("~output_topic", "/tianracer/ackermann_cmd"))
        self.collision_topic = str(rospy.get_param("~collision_topic", "/tianracer/collision"))
        self.collision_required = bool(rospy.get_param("~collision_required", False))
        self.command_timeout = float(rospy.get_param("~command_timeout_s", 0.20))
        self.collision_timeout = float(rospy.get_param("~collision_timeout_s", 0.50))
        self.rate = float(rospy.get_param("~publish_rate_hz", 20.0))
        self._lock = threading.Lock()
        self._enabled = bool(rospy.get_param("~start_enabled", False))
        self._command = None
        self._command_time = float("-inf")
        self._collision = False
        self._collision_time = float("-inf")
        self.pub = rospy.Publisher(self.output_topic, AckermannDrive, queue_size=1)
        self.sub = rospy.Subscriber(self.input_topic, AckermannDriveStamped, self._command_cb, queue_size=1)
        self.collision_sub = None
        if self.collision_required:
            self._collision = True
            self.collision_sub = rospy.Subscriber(
                self.collision_topic, Bool, self._collision_cb, queue_size=1
            )
        self.service = rospy.Service("~set_enabled", SetBool, self._enable_cb)
        self.timer = rospy.Timer(rospy.Duration.from_sec(1.0 / self.rate), self._tick)
        rospy.on_shutdown(self._stop)
        self._stop()
        rospy.logwarn(
            "MPCC command gate ready output=%s enabled=%s",
            self.output_topic,
            self._enabled,
        )

    def _command_cb(self, message: AckermannDriveStamped) -> None:
        command = AckermannDrive()
        command.steering_angle = message.drive.steering_angle
        command.speed = message.drive.speed
        with self._lock:
            self._command = command
            self._command_time = time.monotonic()

    def _collision_cb(self, message: Bool) -> None:
        with self._lock:
            self._collision = bool(message.data)
            self._collision_time = time.monotonic()

    def _enable_cb(self, request: SetBool.Request) -> SetBoolResponse:
        with self._lock:
            now = time.monotonic()
            safe = (
                self._command is not None
                and now - self._command_time <= self.command_timeout
                and (
                    not self.collision_required
                    or (
                        now - self._collision_time <= self.collision_timeout
                        and not self._collision
                    )
                )
            )
            if request.data and not safe:
                requirement = (
                    "fresh command and clear collision state"
                    if self.collision_required
                    else "fresh command"
                )
                return SetBoolResponse(False, requirement + " required")
            self._enabled = bool(request.data)
        if not request.data:
            self._stop()
        return SetBoolResponse(True, "enabled" if request.data else "disabled")

    def _stop(self) -> None:
        self.pub.publish(AckermannDrive())

    def _tick(self, _event) -> None:
        with self._lock:
            now = time.monotonic()
            safe = (
                self._enabled
                and self._command is not None
                and now - self._command_time <= self.command_timeout
                and (
                    not self.collision_required
                    or (
                        now - self._collision_time <= self.collision_timeout
                        and not self._collision
                    )
                )
            )
            command = self._command if safe else None
        self.pub.publish(command if command is not None else AckermannDrive())


def main() -> None:
    rospy.init_node("f1tenth_mpcc_command_gate")
    CommandGate()
    rospy.spin()


if __name__ == "__main__":
    main()
